pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,554 @@
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTrafficControl(key); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing TrafficControl failed, requeue", "TrafficControl", key)
+	}
+	return true
+}
+
+func (c *Controller) createTcState(name string) *tcState {
+	c.tcStatesMutex.Lock()
+	defer c.tcStatesMutex.Unlock()
+	state := &tcState{
+		ofPorts: sets.NewInt32(),
+		pods:    sets.NewString(),
+	}
+	c.tcStates[name] = state
+	return state
+}
+
+func (c *Controller) getTcState(name string) (*tcState, bool) {
+	c.tcStatesMutex.RLock()
+	defer c.tcStatesMutex.RUnlock()
+	state, exists := c.tcStates[name]
+	return state, exists
+}
+
+func (c *Controller) deleteTcState(name string) {
+	c.tcStatesMutex.Lock()
+	defer c.tcStatesMutex.Unlock()
+	delete(c.tcStates, name)
+	return
+}
+
+func (c *Controller) filterPods(appliedTo *v1alpha2.AppliedTo) ([]*v1.Pod, error) {
+	if appliedTo == nil {
+		return c.podLister.List(labels.Everything())
+	}
+	var podSelector, nsSelector labels.Selector
+	var err error
+
+	if appliedTo.PodSelector == nil {
+		podSelector = labels.Everything()
+	} else {
+		if podSelector, err = metav1.LabelSelectorAsSelector(appliedTo.PodSelector); err != nil {
+			return nil, err
+		}
+	}
+
+	if appliedTo.NamespaceSelector != nil {
+		var namespaces []*v1.Namespace
+		if nsSelector, err = metav1.LabelSelectorAsSelector(appliedTo.NamespaceSelector); err != nil {
+			return nil, err
+		}
+		if namespaces, err = c.namespaceLister.List(nsSelector); err != nil {
+			return nil, err
+		}
+		var allPods []*v1.Pod
+		for _, ns := range namespaces {
+			pods, err := c.podLister.Pods(ns.Name).List(podSelector)
+			if err != nil {
+				return nil, err
+			}
+			allPods = append(allPods, pods...)
+		}
+		return allPods, nil
+	}
+	return c.podLister.List(podSelector)
+}
+
+func genName(name string) string {
+	return uuid.NewV5(uuidNamespace, name).String()
+}
+
+func genPortNameUDPTunnel(tunnel *v1alpha2.UDPTunnel) string {
+	names := []string{tunnel.RemoteIP}
+	if tunnel.DestinationPort != nil {
+		names = append(names, strconv.Itoa(int(*tunnel.DestinationPort)))
```

Comment:
I think it should use DestinationPort=4789 to generate port name when it's nil, otherwise `{remoteIP: 1.1.1.1, destinationPort: 4789}` would conflict with `{remoteIP: 1.1.1.1}` in OVS.

Reply:
Done.